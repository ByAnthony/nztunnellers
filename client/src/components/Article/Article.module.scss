@use 'src/index.scss';

.container {
  @include index.component-margin;

  margin-top: index.$l-margin;

  h1 {
    margin-top: 0;
    margin-bottom: 0;
  }

  p {
    margin: index.$m-margin 0;
  }

  @include index.tablet {
    margin-top: index.$xl-margin;
  }

  @include index.desktop {
    margin-top: index.$xxl-margin;
  }
}

.links {
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  padding: index.$base-margin 0;
  gap: index.$s-margin;

  @include index.tablet {
    padding: index.$base-margin 0;
  }

  @include index.desktop {
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-between;
    padding: index.$l-margin 0;
  }
}

.link-container {
  height: 350px;
  width: 15rem;
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;

  @include index.featured-button-link;

  p {
    text-align: left;
  }

  .link-button {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: flex-start;
    align-content: space-between;
    padding: index.$m-margin;
    font-size: 1.5rem;
    color: index.$secondary-color;

    p {
      margin: 0 0 5px 0;
      font-size: 1.1rem;
      color: index.$secondary-color;

      @include index.desktop {
        font-size: 1.2rem;
      }
    }

    @include index.tablet {
      padding: index.$m-margin;
      font-size: 1.75rem;
    }

    @include index.desktop {
      padding: index.$padding-card;
      font-size: 1.8rem;
    }
  }
}
