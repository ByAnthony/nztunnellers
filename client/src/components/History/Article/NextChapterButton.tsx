import { Next } from '../../../types/article';
import STYLES from '../History.module.scss';

type Props = {
    next: Next | null;
};

export function NextChapterButton({ next }: Props) {
  if (next) {
    return (
      <div className={STYLES['button-chapter-container']}>
        <a
          href={`/history/${next.url}`}
          className={STYLES['button-chapter']}
          aria-label={`Go to Chapter ${next.chapter}: ${next.title}`}
        >
          <div>
            <p>{`Chapter ${next.chapter}`}</p>
            <span>{next.title}</span>
          </div>
          <div className={STYLES.arrow}>&rarr;</div>
        </a>
      </div>
    );
  }
  return null;
}
