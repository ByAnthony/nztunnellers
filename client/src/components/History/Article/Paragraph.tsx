import { Section } from '../../../types/article';
import STYLES from '../History.module.scss';

type Props = {
    section: Section;
}

export const formatText = (text: string) => {
  const paragraphs = text.split('\\n\\n');
  return paragraphs.map((paragraph) => (
    <p key={paragraphs.indexOf(paragraph)}>
      {paragraph
        .replace(/--/g, '\u00A0')
        .split(/(\*.+?\*)|(\[.+?\))/g)
        .map((segment) => {
          if (segment && segment.startsWith('*') && segment.endsWith('*')) {
            const italicText = segment.slice(1, -1);
            return <em key={paragraph.indexOf(segment)}>{italicText}</em>;
          }
          if (segment && segment.startsWith('[') && segment.endsWith(')')) {
            const linkText = segment.slice(1, -1);
            const [label, url] = linkText.split('](');
            if (url.includes('footnote')) {
              return (
                <a key={paragraph.indexOf(segment)} href={url} id={`reference_${label}`}>
                  [
                  {label}
                  ]
                </a>
              );
            }
            const number = label.slice(0, -1);
            return (
              <a key={paragraph.indexOf(segment)} href={url} id={`footnote_${number}`}>
                {label}
              </a>
            );
          }
          return <span key={paragraph.indexOf(segment)}>{segment}</span>;
        })}
    </p>
  ));
};

export function Paragraph({ section }: Props) {
  return (
    <div className={STYLES.paragraph}>
      <h2>{section.title}</h2>
      {formatText(section.text)}
    </div>
  );
}
