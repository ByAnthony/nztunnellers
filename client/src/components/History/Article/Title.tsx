import STYLES from '../History.module.scss';

type Props = {
  title: string;
  chapter: number;
}

export function Title({ title, chapter }: Props) {
  const [titleLine1, titleLine2] = title.split('\\');

  return (
    <div className={STYLES['chapter-title']}>
      <h1>
        <span className={STYLES['sub-title']}>{titleLine1}</span>
        <span className={STYLES.title}>{titleLine2}</span>
      </h1>
      <div className={STYLES['chapter-number']}>{chapter}</div>
    </div>
  );
}
