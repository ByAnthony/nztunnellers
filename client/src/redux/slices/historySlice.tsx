import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { Article, Articles } from '../../types/article';

export const historyApi = createApi({
  reducerPath: 'historyApi',
  baseQuery: fetchBaseQuery({
    baseUrl: 'http://localhost:5000/',
  }),
  endpoints: (builder) => ({
    getAllHistoryArticleLink: builder.query<Array<Articles>, void>({
      query: () => ({
        url: '/',
      }),
    }),
    getHistoryArticleById: builder.query<Article, string>({
      query: (id: string) => `history/${id}`,
    }),
  }),
});

export const {
  useGetAllHistoryArticleLinkQuery,
  useGetHistoryArticleByIdQuery,
} = historyApi;
