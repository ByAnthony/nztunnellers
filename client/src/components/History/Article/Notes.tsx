import STYLES from '../History.module.scss';
import { formatText } from './Paragraph';

type Props = {
    notes: string;
};

export function Notes({ notes }: Props) {
  return (
    <div className={STYLES.notes}>
      <h3>Notes</h3>
      {formatText(notes)}
    </div>
  );
}
