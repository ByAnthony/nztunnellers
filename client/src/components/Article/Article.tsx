import { useParams } from 'react-router-dom';
import { useGetHistoryArticleByIdQuery } from '../../redux/slices/rollSlice';

// import { invariant } from '../../utils/invariant';
import { today } from '../../utils/date';

import { Content } from './Content/Content';
import { Footer } from '../Footer/Footer';
import { HowToCite } from '../HowToCite/HowToCite';
import { Menu } from '../Menu/Menu';
import { NextChapterButton } from './NextChapterButton/NextChapterButton';
import { Notes } from './Notes/Notes';
import { Title } from '../Title/Title';
import { TopImage } from './TopImage/TopImage';

import STYLES from './Article.module.scss';

export function Article() {
  const { id = '' } = useParams();

  const {
    data, error, isLoading, isSuccess,
  } = useGetHistoryArticleByIdQuery(id);

  if (data) {
    return (
      <>
        {error && (
        <div className={STYLES.profile}>
          <p>An error occured</p>
        </div>
        )}
        <Menu />
        { isLoading }
        { isSuccess && (
        <div className={STYLES.container}>
          <div className={STYLES.header}>
            <div className={STYLES.link}>
              <a href="/#history">History</a>
            </div>
            <Title title={data.title} subTitle={data.chapter} />
          </div>
          <TopImage image={data.image[0]} />
          <Content imageList={data.image.slice(1)} sectionList={data.section} />
          <NextChapterButton chapter={data.next} />
          <Notes notes={data.notes} />
          <HowToCite title={data.title} today={today} />
        </div>
        )}
        <Footer />
      </>
    );
  }
  return null;
}
