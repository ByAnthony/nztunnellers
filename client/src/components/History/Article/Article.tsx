import { useParams } from 'react-router-dom';
import { Footer } from '../../Footer/Footer';
import { Menu } from '../../Menu/Menu';
import { Title } from './Title/Title';
import { today } from '../../../utils/date-utils';
import { useGetHistoryArticleByIdQuery } from '../../../redux/slices/rollSlice';
import STYLES from './Article.module.scss';
import { HowToCite } from './HowToCite/HowToCite';
import { Notes } from './Notes/Notes';
import { NextChapterButton } from './NextChapterButton/NextChapterButton';
import { TopImage } from './TopImage/TopImage';
import { Content } from './Content/Content';

export function Article() {
  const { id } = useParams();
  const {
    data, error, isLoading, isSuccess,
  } = useGetHistoryArticleByIdQuery(id!);

  if (data) {
    return (
      <>
        {error && (
        <div className={STYLES.profile}>
          <p>An error occured</p>
        </div>
        )}
        <Menu />
        { isLoading }
        { isSuccess && (
        <div className={STYLES.container}>
          <div className={STYLES.header}>
            <div className={STYLES.link}>
              <a href="/#history">History</a>
            </div>
            <Title title={data.title} chapter={data.chapter} />
          </div>
          <TopImage image={data.image[0]} />
          <Content imageList={data.image.slice(1)} sectionList={data.section} />
          <NextChapterButton chapter={data.next} />
          <Notes notes={data.notes} />
          <HowToCite title={data.title} today={today} />
        </div>
        )}
        <Footer />
      </>
    );
  }
  return null;
}
