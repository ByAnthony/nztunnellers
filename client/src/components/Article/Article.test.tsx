import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';

import { useGetHistoryArticleByIdQuery } from '../../redux/slices/historySlice';
import { mockArticle } from '../../utils/mocks/mockArticle';

import { Article } from './Article';

jest.mock('../../redux/slices/historySlice', () => ({
  useGetHistoryArticleByIdQuery: jest.fn(),
}));

jest.mock('../../utils/date', () => ({
  today: new Date('2023-05-04'),
}));

test('renders profile when data is available', () => {
  (useGetHistoryArticleByIdQuery as jest.Mock).mockReturnValue({
    data: mockArticle,
    error: null,
    isLoading: false,
    isSuccess: true,
  });

  const { asFragment } = render(<Article />);

  expect(asFragment()).toMatchSnapshot();
});

test('does not render profile when data is undefined', () => {
  (useGetHistoryArticleByIdQuery as jest.Mock).mockReturnValue({
    data: undefined,
    error: null,
    isLoading: false,
    isSuccess: true,
  });

  const { container } = render(<Article />);

  expect(container).toBeEmptyDOMElement();
});

test('should render error page when error', () => {
  (useGetHistoryArticleByIdQuery as jest.Mock).mockReturnValue({
    data: {},
    error: true,
    isLoading: false,
    isSuccess: false,
  });

  render(<Article />);
  const error = screen.getByText('An error occured');

  expect(error).toBeInTheDocument();
});
