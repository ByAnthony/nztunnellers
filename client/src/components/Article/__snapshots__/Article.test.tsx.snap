// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders profile when data is available 1`] = `
<DocumentFragment>
  <div
    class="menu "
    data-testid="menu"
  >
    <a
      aria-label="Go to the Homepage"
      class="logo"
      href="/"
    >
      <img
        alt=""
        src="/nzt_logo.png"
      />
    </a>
    <div
      class="search-form"
    >
      <form>
        <input
          id="search"
          name="search"
          placeholder="Search for a Tunneller"
          type="text"
        />
      </form>
      <button
        aria-label="Search"
        class="search-form-button"
        type="submit"
      >
        <img
          alt=""
          height="18"
          src="/searching_hover.png"
        />
      </button>
    </div>
  </div>
  <div
    class="container"
  >
    <div
      class="header"
    >
      <div
        class="link"
      >
        <a
          href="/#history"
        >
          History
        </a>
      </div>
      <div
        class="main-title"
      >
        <h1>
          <span
            class="title-line-1"
          >
            My Awesome
          </span>
          <span
            class="title-line-2"
          >
            Article Title
          </span>
        </h1>
      </div>
      <h2
        class="title-line-3"
      >
        Chapter 1
      </h2>
    </div>
    <div
      class="image-container"
    >
      <img
        alt="Accessible alt text"
        class="image"
        src="/images/history/img-123.png"
      />
    </div>
    <div>
      <div
        class="paragraph"
      >
        <h2>
          Section Title
        </h2>
        <p>
          <span>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea 
          </span>
          <em>
            commodo consequat
          </em>
          <span />
          <span>
            . Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
          </span>
        </p>
        <p>
          <span>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat
          </span>
          <span />
          <a
            href="#footnote_1"
            id="reference_1"
          >
            [1]
          </a>
          <span>
            . Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
          </span>
        </p>
        <p>
          <span>
            (6 July 2023)
          </span>
        </p>
      </div>
      <div
        class="image-container"
      >
        <img
          alt="Accessible alt text"
          class="image"
          src="/images/history/img-123.png"
        />
      </div>
      <div
        class="image-legend"
      >
        <div
          class="title"
        >
          An Amazing Photo
        </div>
        <div
          class="captions"
        >
          The French Photographer Doisneau
        </div>
        <div
          class="reference"
        >
          Where the photo is preserved
        </div>
      </div>
    </div>
    <div>
      <div
        class="paragraph"
      >
        <h2>
          Section Title
        </h2>
        <p>
          <span>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea 
          </span>
          <em>
            commodo consequat
          </em>
          <span />
          <span>
            . Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
          </span>
        </p>
        <p>
          <span>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit,sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat
          </span>
          <span />
          <a
            href="#footnote_1"
            id="reference_1"
          >
            [1]
          </a>
          <span>
            . Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
          </span>
        </p>
        <p>
          <span>
            (6 July 2023)
          </span>
        </p>
      </div>
    </div>
    <div
      class="button-chapter-container"
    >
      <a
        aria-label="Go to Chapter 3: Next Chapter"
        class="button-chapter"
        href="/history/my-path-to-next-chapter/"
      >
        <div>
          <p
            class="chapter"
          >
            Chapter 3
          </p>
          <span>
            Next Chapter
          </span>
        </div>
        <div
          class="arrow"
        >
          →
        </div>
      </a>
    </div>
    <div
      class="notes"
    >
      <h3>
        Notes
      </h3>
      <p>
        <span />
        <span />
        <a
          href="#reference_1"
          id="footnote_1"
        >
          1.
        </a>
        <span>
           Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
        </span>
      </p>
      <p>
        <span />
        <span />
        <a
          href="#reference_2"
          id="footnote_2"
        >
          2.
        </a>
        <span>
           Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
        </span>
      </p>
    </div>
    <div
      class="howtocite"
    >
      <h3>
        How to cite this page
      </h3>
      <p>
        Anthony Byledbal, “
        <span>
          My Awesome Article Title
        </span>
        “,
        <em>
           New Zealand Tunnellers Website
        </em>
        , 2023 (2009), Accessed: 4 May 2023. 
        <span>
          URL: www.
          <wbr />
          nztunnellers
          <wbr />
          .com/
          <wbr />
          history/
          <wbr />
          my-awesome-article-title.
        </span>
      </p>
    </div>
  </div>
  <div
    class="footer"
  >
    <div
      class="map"
    >
      <div
        class="links"
      >
        <div
          class="map-link"
        >
          <a
            href="/#history"
          >
            History
          </a>
        </div>
        <div
          class="map-link"
        >
          <a
            href="/tunnellers"
          >
            Tunnellers
          </a>
        </div>
        <div
          class="map-link"
        >
          <a
            href="/about-us"
          >
            About Us
          </a>
        </div>
      </div>
      <button
        aria-label="Go back to the top of the page"
        class="scroll-top"
        type="button"
      >
        ↑
      </button>
    </div>
    <div
      class="support"
    >
      <div>
        <a
          aria-label="Go to The Artois University website"
          href="https://www.univ-artois.fr/artois-university"
        >
          <img
            alt="Artois University homepage"
            src="/images/support/logo-univ-artois-blanc_0.png"
          />
        </a>
      </div>
      <div>
        <a
          aria-label="Go to The Institute for Strategic Research website"
          href="https://www.irsem.fr/en/"
        >
          <img
            alt="Institute for Strategic Research homepage"
            src="/images/support/irsem-white.png"
          />
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
